sympy/solvers/simplex.py

```diff
@@ -0,0 +1,821 @@
+    xx = list(ordered(syms)) + aux
+    A, B = linear_eq_to_matrix(np, xx)
+    C, D = linear_eq_to_matrix([F], xx)
+    return A, B, C, D, r, xx, aux
+
+
+def _lp(min_max, f, constr):
+    """Return the optimization (min or max) of ``f`` with the given
+    constraints. Unless constrained, variables will assume only
+    nonnegative values in the solution. A constraint like ``x <= 2``
+    will permit ``x`` to have negative values while ``x < oo`` will
+    allow it to have any real value.
+
+    If `min_max` is 'max' then the results corresponding to the
+    maximization of ``f`` will be returned, else the minimization.
+    The constraints can be given as Le, Ge or Eq expressions.
+
+    Examples
+    ========
+
+    >>> from sympy.solvers.simplex import _lp as lp
+    >>> from sympy import symbols, Eq, oo
+    >>> from sympy.abc import x, y
+    >>> x1, x2, x3, x4 = symbols('x1:5')
+    >>> f = 5*x2 + x3 + 4*x4 - x1
+    >>> c = [x1 + 5 >= 5*x2 + 2*x3 + 5*x4,
+    ...      Eq(3*x2 + x4, 2),
+    ...      Eq(-x1 + x3 + 2*x4, 1)]
+    >>> lp(min, f, c)
+    (9/2, {x1: 0, x2: 1/2, x3: 0, x4: 1/2})
+
+    By passing max, the maximum value for f under the constraints
+    is returned if possible:
+
+    >>> lp(max, f, c + [x3 < oo])  # x3 can have any value
+    (5, {x1: 1, x2: 0, x3: -2, x4: 2})
+
+    Symbols not in the objective function will not be reported
+    and are treated as 0:
+
+    >>> lp(min, x, [y + x >= -3])  # same as lp(min, x, [x >= -3])
+    (-3, {x: -3})
+    """
+    A, B, C, D, r, xx, aux = _lp_matrices(f, constr)
+
+    how = str(min_max).lower()
+    if 'max' in how:
+        # _simplex minimizes for Ax <= B so we
+        # have to change the sign of the function
+        # and negate the optimal value returned
+        _o, p, d = _simplex(A, B, -C, -D)
+        o = -_o
+    elif 'min' in how:
+        o, p, d = _simplex(A, B, C, D)
+    else:
+        raise ValueError('expecting min or max')
+
+    # restore original variables and remove aux from p
+    p = dict(zip(xx, p))
+    if r:  # p has original symbols and auxilliary symbols
+        # if r has x: x - z1 use values from p to update
+        r = {k: v.xreplace(p) for k, v in r.items()}
+        # then use the actual value of x (= x - z1) in p
+        p.update(r)
+        # and keep only non-aux variables
+        p = {k: v for k, v in p.items() if k not in aux}
+
+    # not returning dual since there may be extra constraints
+    # when a variable has finite bounds
+    return o, p
+
+
+def _lp_args(min_max, *args):
+    """return standard input for lp from how and args. The
+    method ``how`` is interpreted as being either min or max,
+    and args (up to 4) are interpreted and returned as an
+    object function and list of appropriate inequalities.
+
+    1 arg: Matrix/list
+    2 args: Expr, list of constraints
+    3 args: 3 lists/matrices: A, B, C, D=0
+    4 args: 4 lists/matrices: A, B, C, D
+
+    return: ('max' or 'min', objective, constraints)
```

Comment:
So the returned objective function is an expression and constraints contain a list of Ge, Le, Eq?

It was my understanding that the purpose of being able to handle matrices was for efficiency reasons.  Wouldn't converting large matrices into SymPy expressions and then back into matrices be slow?

Why not have _lp_args convert everything to a standard matrix form?

Reply:
I'll look at refactoring. It might take some time given the current schedule.

Reply:
Now that I am not solving equalities, I could stay in "matrix mode" in `linprog`. But that might take another week to get time enough to finish it.